### PR TITLE
Develop causal

### PR DESCRIPTION
**Pull request by bblodfon**
Updated the following psimi modules to include causality (support of the MITAB 2.8 format):

- calimocho
- calimocho-tab
- calimocho-solr
- psidev.psi.mi.tab (psimitab)

### CODE REVIEW

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/AnnotationFieldConverterTest.java [hunk 1]
```diff
@@ -58,81 +59,65 @@ protected void setUp() throws Exception {
...
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab27 = MitabDocumentDefinitionFactory.mitab27();
-        DefaultRowReader rowReader_mitab27 = new DefaultRowReader( documentDefinition_mitab27 );
-        rowList_mitab27 = rowReader_mitab27.read(File.class.getResourceAsStream("/samples/sampleFileMitab27.txt"));
-
         ColumnBasedDocumentDefinition documentDefinition_mitab26 = MitabDocumentDefinitionFactory.mitab26();
         DefaultRowReader rowReader_mitab26 = new DefaultRowReader( documentDefinition_mitab26 );
         rowList_mitab26 = rowReader_mitab26.read(File.class.getResourceAsStream("/samples/sampleFileMitab26.txt"));
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab25 = MitabDocumentDefinitionFactory.mitab25();
```
Comment (noedelta) [inline calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/AnnotationFieldConverterTest.java]: I have the impression that the test for mitab25 could be still valid, and maybe we should keep it. What do you think?

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/DateFieldConverterTest.java [hunk 2]
```diff
@@ -58,30 +59,32 @@ protected void setUp() throws Exception {
...
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab27 = MitabDocumentDefinitionFactory.mitab27();
-        DefaultRowReader rowReader_mitab27 = new DefaultRowReader( documentDefinition_mitab27 );
-        rowList_mitab27 = rowReader_mitab27.read(File.class.getResourceAsStream("/samples/sampleFileMitab27.txt"));
-
         ColumnBasedDocumentDefinition documentDefinition_mitab26 = MitabDocumentDefinitionFactory.mitab26();
         DefaultRowReader rowReader_mitab26 = new DefaultRowReader( documentDefinition_mitab26 );
         rowList_mitab26 = rowReader_mitab26.read(File.class.getResourceAsStream("/samples/sampleFileMitab26.txt"));
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab25 = MitabDocumentDefinitionFactory.mitab25();
```
Comment (noedelta) [inline calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/DateFieldConverterTest.java]: Same than before, maybe we could keep the test for column definition of the 2.5

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/TextToBooleanFieldConverterTest.java [hunk 3]
```diff
@@ -58,30 +58,32 @@ protected void setUp() throws Exception {
...
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab27 = MitabDocumentDefinitionFactory.mitab27();
-        DefaultRowReader rowReader_mitab27 = new DefaultRowReader( documentDefinition_mitab27 );
-        rowList_mitab27 = rowReader_mitab27.read(File.class.getResourceAsStream("/samples/sampleFileMitab27.txt"));
-
         ColumnBasedDocumentDefinition documentDefinition_mitab26 = MitabDocumentDefinitionFactory.mitab26();
         DefaultRowReader rowReader_mitab26 = new DefaultRowReader( documentDefinition_mitab26 );
         rowList_mitab26 = rowReader_mitab26.read(File.class.getResourceAsStream("/samples/sampleFileMitab26.txt"));
 
-        ColumnBasedDocumentDefinition documentDefinition_mitab25 = MitabDocumentDefinitionFactory.mitab25();
```
Comment (noedelta) [inline calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/TextToBooleanFieldConverterTest.java]: The same

File: calimocho-tab/src/test/java/org/hupo/psi/calimocho/tab/io/DefaultRowReaderTest.java [hunk 4]
```diff
@@ -137,6 +266,141 @@ public void readStream() throws Exception {
...
+        final String value = pmethodBField.get( CalimochoKeys.VALUE );
+        final String text = pmethodBField.get( CalimochoKeys.TEXT );
+
+        Assert.assertEquals("psi-mi", key);
+        Assert.assertEquals("MI:0102", value);
+        Assert.assertEquals("sequence tag identification", text);
+    }
+
+    @Test
+    public void reamMitab28Stream() throws Exception {
```
Comment (noedelta) [inline calimocho-tab/src/test/java/org/hupo/psi/calimocho/tab/io/DefaultRowReaderTest.java]: Typo ream -> read

File: psimitab-parser/src/main/java/psidev/psi/mi/tab/model/Interactor.java [hunk 5]
```diff
@@ -367,9 +372,24 @@ public void setParticipantIdentificationMethods(List<CrossReference> participant
...
+	 *
+	 * @return Value for property 'biological effect(s)'.
+	 */
+	public void setBiologicalEffects(List<CrossReference> biologicalEffects) {
+		this.biologicalEffects = biologicalEffects;
+	}
 
 	public boolean isEmpty() {
-		//We don not want to have a empty interactor, we prefer a null interactor
+		//We don't not want to have a empty interactor, we prefer a null interactor
```
Comment (noedelta) [inline psimitab-parser/src/main/java/psidev/psi/mi/tab/model/Interactor.java]: Still the not is there "We don't want"

File: psimitab-parser/src/main/java/psidev/psi/mi/tab/model/ParameterImpl.java [hunk 6]
```diff
@@ -217,10 +219,13 @@ public void setValue(String value) {
                     case 3:
                         if (!matcher.group(3).isEmpty()) {
                             exponent = Integer.parseInt(matcher.group(3));
+                            userDefinedExponent = true;
                         }
                     case 2:
                         if (!matcher.group(2).isEmpty()) {
                             base = Integer.parseInt(matcher.group(2));
+                            userDefinedBase = true;
+                            if (base == 0) throw new IllegalArgumentException("No point in setting base to zero");
```
Comment (noedelta) [inline psimitab-parser/src/main/java/psidev/psi/mi/tab/model/ParameterImpl.java]: Can we update the message to something like: "Zero is not allowed as a value for base" ?

File: psimitab-parser/src/main/java/psidev/psi/mi/tab/model/ParameterImpl.java [hunk 7]
```diff
@@ -186,9 +181,16 @@ public void setExponent(Integer exponent) {
     public String getValue() {
         this.value = String.valueOf(factor);
 
-		if (exponent != 0 || base != 10) {
-			value = factor + "x" + base + "^" + exponent;
-		}
+        if (base == 1 || (!userDefinedExponent && !userDefinedBase)) {
```
Comment (noedelta) [inline psimitab-parser/src/main/java/psidev/psi/mi/tab/model/ParameterImpl.java]: Noe: note for myself

File: psimitab-parser/src/test/java/psidev/psi/mi/tab/PsimiTabWriterTest.java [hunk 8]
```diff
@@ -44,6 +44,16 @@ private int lineCount(File file) throws IOException {
         return count;
     }
 
+    private void printlines(File file) throws IOException {
```
Comment (noedelta) [inline psimitab-parser/src/test/java/psidev/psi/mi/tab/PsimiTabWriterTest.java]: Typo printlines -> printLines

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/AnnotationFieldConverterTest.java (hunk 1 above)
Comment (bblodfon) [reply calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/AnnotationFieldConverterTest.java]: There were no annotation fields in MITAB 2.5 that's why I removed it.

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/DateFieldConverterTest.java (hunk 2 above)
Comment (bblodfon) [reply calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/DateFieldConverterTest.java]: No dates in MITAB 2.5!

File: calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/TextToBooleanFieldConverterTest.java (hunk 3 above)
Comment (bblodfon) [reply calimocho-solr/src/test/java/psidev/psi/mi/calimocho/solr/converter/TextToBooleanFieldConverterTest.java]: No interaction parameters in MITAB 2.5!

**Pull request merged**